### PR TITLE
[inductor] custom do_bench_gpu to reduce max-autotune overhead

### PR DESCRIPTION
**Pull request by nmacchioni**
Differential Revision: D58012164

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/runtime/runtime_utils.py [hunk 1]
```diff
@@ -81,35 +80,62 @@ def do_bench(fn, fn_args, fn_kwargs, **kwargs):
...
+    ]
+    for start_event, end_event in event_pairs:
+        cache.zero_()
+        start_event.record()
+        fn()
+        end_event.record()
+    torch.cuda.synchronize()
+    estimate_ms = min(
+        [event_pair[0].elapsed_time(event_pair[1]) for event_pair in event_pairs]
+    )
```
Comment (eellison) [inline torch/_inductor/runtime/runtime_utils.py:112]: From https://github.com/triton-lang/triton/pull/3943#issuecomment-2128027979: 

> I agree that min is much more stable than mean, and that's probably what I'd do if I were starting from scratch. But this biases you towards algorithms which perform better when the GPU is cold, and I believe we have seen cases where that isn't globally ideal. Just adding to the complexity...

Could we maintain current behavior of median and consider switching in a future pr ?

File: torch/_inductor/runtime/runtime_utils.py [hunk 2]
```diff
@@ -81,35 +80,62 @@ def do_bench(fn, fn_args, fn_kwargs, **kwargs):
         return do_bench_gpu(lambda: fn(*fn_args, **fn_kwargs), **kwargs)
 
 
-def do_bench_gpu(*args, **kwargs):
+def do_bench_gpu(fn, warmup=25, rep=100, fast_flush=True):
```
Comment (eellison) [inline torch/_inductor/runtime/runtime_utils.py:83]: this removes some arguments, but as far as i can tell the removed args are not used in pytorch.

File: torch/_inductor/runtime/runtime_utils.py [hunk 3]
```diff
@@ -81,35 +80,62 @@ def do_bench(fn, fn_args, fn_kwargs, **kwargs):
...
-        # Maybe we should have own implementation of this function?
-        return triton_do_bench, (
-            "quantiles"
-            if inspect.signature(triton_do_bench).parameters.get("quantiles")
-            is not None
-            else "percentiles"
+    def get_cache_size():
+        device = torch.cuda.current_device()
+        properties = torch.cuda.get_device_properties(device)
+        return properties.l2CacheSize
```
Comment (eellison) [inline torch/_inductor/runtime/runtime_utils.py:88]: >  (A problem with flushing exactly the L2 cache size is that you don't actually know that this clears the cache. For example H100 has two separate L2s, and I don't know how they work. But again, touching 4MB vs 256MB should not be a substantial difference, certainly not >1ms on H100, so if you're observing that, it seems like something is wrong.)

Should we double on h100 ?

File: torch/_inductor/runtime/runtime_utils.py [hunk 4]
```diff
@@ -81,35 +80,62 @@ def do_bench(fn, fn_args, fn_kwargs, **kwargs):
...
+
+    if fast_flush:
```
Comment (eellison) [inline torch/_inductor/runtime/runtime_utils.py:90]: We could just ignore fast_flush, I don't know when youd ever want to do slow flush haha

File: torch/_inductor/runtime/runtime_utils.py [hunk 5]
```diff
@@ -81,35 +80,62 @@ def do_bench(fn, fn_args, fn_kwargs, **kwargs):
...
+    for start_event, end_event in event_pairs:
+        cache.zero_()
+        start_event.record()
+        fn()
+        end_event.record()
+    torch.cuda.synchronize()
+
+    timing = min(
+        [event_pair[0].elapsed_time(event_pair[1]) for event_pair in event_pairs]
+    )
```
Comment (eellison) [inline torch/_inductor/runtime/runtime_utils.py:137]: From https://github.com/triton-lang/triton/pull/3943#issuecomment-2128027979: 

> I agree that min is much more stable than mean, and that's probably what I'd do if I were starting from scratch. But this biases you towards algorithms which perform better when the GPU is cold, and I believe we have seen cases where that isn't globally ideal. Just adding to the complexity...

Could we maintain current behavior of median and consider switching in a future pr ?

Or continue to expose `return_mode`

**Review by eellison — COMMENTED** [review]
Looks great! have a few comments

File: torch/_inductor/runtime/runtime_utils.py (hunk 2 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/runtime_utils.py:83]: Yeah, I did a quick search and the removed arguments (grad_to_none, quantiles, return_mode) were all previously unused (grad_to_none) or now unused (quantiles, return_mode). For code simplicity I opted to completely remove them.

File: torch/_inductor/runtime/runtime_utils.py (hunk 3 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/runtime_utils.py:88]: As far as I know, H100 has a single L2 cache; I've done some research and nothing came up mentioning two L2 caches, and experimentally 50MB flushes showed equivalent performance (whereas 49MB did not produce the correct results). I'm inclined to believe this was a typo on their part, probably meant B100. In that case, I have no idea how Nvidia will handle this. Should their CUDA device properties return the total L2 cache size? Probably. I won't know until we get B100s to test. In the meantime I don't think this should be a big issue due to the incredibly limited quantity of B100s out in the wild. My opinion here is that we just land now and check when we have access later, but I could see why this isn't amazing either.

File: torch/_inductor/runtime/runtime_utils.py (hunk 5 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/runtime_utils.py:137]: Sure, good idea to seperate that out into a future PR.

More on that idea though, I think both min and quantile are inherently flawed in their own ways. This all loops back to the main issue that autotuning is really hard, and you can find problems which pretty much every way of doing it. 

Yes, min is can have problems when running on a cold GPU. But really, isn't that what the warmups are supposed to take care of? How many warmups are enough to get the GPU warm without cooking it? It is very hard to get this even close to right.

Any sort of quantile also has its own problems. The quantiles can get pretty heavily skewed if anything disturbs the GPU for a portion of the benchmarking; min is not as sensitive to this issue.

I'll split this off into a separate PR like you mentioned; it really has no bearing on compile time overhead, and is therefore not so relevant to this PR.

File: torch/_inductor/runtime/runtime_utils.py (hunk 4 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/runtime_utils.py:90]: Yeah I'm not particularly sure; I think that would be a good followup PR to remove fast_flush references but I'd like to keep this PR as clean as possible for now.

**Comment by nmacchioni** [comment]
Reverted to original timing method of quantiles/return_mode

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**